src/main/java/org/folio/rest/impl/InventoryHelper.java

```diff
@@ -0,0 +1,429 @@
+                 identifier.put("value", pId.getProductId());
+                 return identifier;
+               })
+               .collect(toList());
+      instance.put("identifiers", new JsonArray(identifiers));
+    }
+    return instance;
+  }
+
+  /**
+   * Search for items which might be already created for the PO line
+   * @param compPOL PO line to retrieve Item Records for
+   * @return future with list of item id's
+   */
+  private CompletableFuture<List<String>> searchForExistingItems(PoLine compPOL) {
+    int expectedCount = calculateInventoryItemsQuantity(compPOL);
+    String endpoint = String.format(LOOKUP_ITEMS_ENDPOINT, compPOL.getId(), expectedCount, lang);
+    return handleGetRequest(endpoint, httpClient, ctx, okapiHeaders, logger)
+      .thenApply(items -> {
+        List<String> itemIds = collectItemIds(items);
+        logger.debug("{} existing items found out of {} for PO Line with '{}' id", itemIds.size(), expectedCount, compPOL.getId());
+        return itemIds;
+      });
+  }
+
+  /**
+   * Validates if the json object contains items and extracts ids or returns empty list
+   * @param itemEntries {@link JsonObject} representing item storage response
+   * @return list of the item ids if any item returned
+   */
+  private List<String> collectItemIds(JsonObject itemEntries) {
+    return Optional.ofNullable(itemEntries.getJsonArray("items"))
+                   .map(items -> items.stream()
+                                      .map(item -> (JsonObject) item)
+                                      .map(this::extractId)
+                                      .collect(toList()))
+                   .orElseGet(Collections::emptyList);
+  }
+
+  private CompletableFuture<List<String>> createMissingItems(PoLine compPOL, String holdingId, int count) {
+    return (count > 0) ? createItemRecords(compPOL, holdingId, count) : completedFuture(Collections.emptyList());
+  }
+
+  /**
+   * Creates Items in the inventory based on the PO line data.
+   *
+   * @param compPOL PO line to create Instance Record for
+   * @param holdingId holding id
+   * @param count expected number of items to create
+   * @return id of newly created Instance Record
+   */
+  private CompletableFuture<List<String>> createItemRecords(PoLine compPOL, String holdingId, int count) {
+    CompletableFuture<List<String>> result = new VertxCompletableFuture<>(ctx);
+    logger.debug("Creating {} items for PO Line with '{}' id", count, compPOL.getId());
+
+    buildItemRecordJsonObject(compPOL, holdingId)
+      .thenCompose(itemData -> createItemRecords(itemData, count))
+      .thenAccept(createdItemIds -> {
+        // In case no items created at all, throw an exception because nothing can be done at this stage
+        if (createdItemIds.isEmpty()) {
+          throw new IllegalStateException(String.format("No items created for PO Line with %s id", compPOL.getId()));
+        } else if (count != createdItemIds.size()) {
```

Comment:
Wouldn't it be better to throw an exception and stop the processing (namely, creating piece records for succeeded items) if not all items are created, than just logging an error and returning succeded ones?
@craigmcnally , what do you think?

Reply:
The operation would need to be retried either way.

That said, on one hand, fast-failing is nice and clean and keeps the code simple.  On the other hand, it's probably cleaner on the storage end to create piece records for the item records that were successfully created.  Is everything in place to fully transition the po_line to Open? No, but at least the records that were created have the appropriate linkage (assuming there aren't also failures while creating the piece records).  

A few other minor factors include:
* the decision we make here could have some affect record metadata...  e.g. creation dates, etc.  I'm not sure it really matters though
* Are the advantages of refactoring this worth the time it's going to take making the changes, getting them reviewed again, holding up other tasks - possibly pushing them into next sprint?

I think I'm in favor of just keeping it as currently implemented.  My decision really comes down to the first and last points I make.